service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java

```diff
@@ -44,23 +51,41 @@ public KubernetesServiceDiscovery(String serviceName, ServiceDiscoveryConfig con
         if (namespace != null && namespace.equalsIgnoreCase("all")) {
             allNamespaces = true;
         }
+        String refreshPeriod = config.parameters().get("refresh-period");
+        try {
+            // TODO: document it
+            this.refreshPeriod = refreshPeriod != null
+                    ? DurationUtils.parseDuration(refreshPeriod)
+                    : ServiceInstancesCache.DEFAULT_REFRESH_INTERVAL;
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("refresh-period for service discovery should be a number, got: " +
+                    refreshPeriod,
+                    e);
+        }
+
         Config properties = new ConfigBuilder(base)
                 .withMasterUrl(masterUrl)
                 .withNamespace(namespace).build();
         client = new DefaultKubernetesClient(properties);
         this.vertx = vertx;
     }
 
-    public Uni<List<ServiceInstance>> blockingGetServiceInstances() {
-        List<Endpoints> endpoints = allNamespaces
-                ? client.endpoints().inAnyNamespace().withField("metadata.name", serviceName).list()
-                        .getItems()
-                : client.endpoints().inNamespace(namespace).withField("metadata.name", serviceName).list().getItems();
-        Uni<List<ServiceInstance>> serviceEntryList = Uni.createFrom().item(map(endpoints));
-        return serviceEntryList; // TODO: logging
+    public Uni<List<ServiceInstance>> getServiceInstances() {
+        if (refreshNotNeed()) {
+            return Uni.createFrom().item(cacheData.getServiceInstances());
+        }
+        Uni<List<ServiceInstance>> serviceInstances = fetchNewServiceInstancesFromK8s();
+        return serviceInstances.onItem()
+                .invoke(services -> this.cacheData = new ServiceInstancesCache(services, LocalDateTime.now()));
     }
 
-    public Uni<List<ServiceInstance>> getServiceInstances() {
+    @Override
```

Comment:
`refreshNotNeeded` is the same for both, right? Also `getServices`?

Reply:
Yes, it's the same. `getServices` also